### PR TITLE
Jet energy resolution smearing with systematic uncertainties split up [102X backport]

### PR DESCRIPTION
**Pull request by ahinzmann**
Backport of #28480

PR description:

Jet energy resolution data/MC scale factors now come with systematic uncertainties broken up into sources. An example is here:
https://github.com/cms-jet/JRDatabase/blob/master/textFiles/Autumn18_V7b_MC/Autumn18_V7b_MC_SF_AK4PFchs.txt
https://github.com/cms-jet/JRDatabase/blob/master/SQLiteFiles/Autumn18_V7b_MC.db
This PR provides an interface to access these uncertainties via:
float getScaleFactor(const JetParameters& parameters, Variation variation = Variation::NOMINAL, std::string uncertaintySource = "") const;
The condition dataformat is modified, however, the persitent part is untoched, so this code also works on old conditions.
PR validation:

It was checked that all uncertainty sources can be accessed with the new interface with the test file runJERsmearingOnMiniAOD.py.
It was checked that the total uncertainty can be accessed as before in old conditions.

Links to discussion in JetMET:
https://indico.cern.ch/event/864045/#2-updates-on-2018-jer-and-disc
https://indico.cern.ch/event/859946/#5-update-on-jer-scale-factors

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ggovi** [comment]
+1

File: CondFormats/JetMETObjects/interface/JetResolutionObject.h [hunk 1]
```diff
@@ -186,6 +186,10 @@ namespace JME {
                         return m_variables.size();
                     }
 
+      const std::vector<std::string>& getParametersName() const { return m_parameters_name; }
+
+      size_t nParameters() const { return m_parameters_name.size(); }
```
Comment (slava77) [inline CondFormats/JetMETObjects/interface/JetResolutionObject.h]: indentation

File: CondFormats/JetMETObjects/src/JetResolutionObject.cc [hunk 2]
```diff
@@ -169,19 +169,36 @@ namespace JME {
         std::string formula_str_lower = m_formula_str;
         std::transform(formula_str_lower.begin(), formula_str_lower.end(), formula_str_lower.begin(), ::tolower);
 
-        if (formula_str_lower == "none")
-            m_formula_str = "";
+    if (formula_str_lower == "none") {
```
Comment (slava77) [inline CondFormats/JetMETObjects/src/JetResolutionObject.cc]: indentation

File: JetMETCorrections/Modules/interface/JetResolution.h [hunk 3]
```diff
@@ -54,7 +54,9 @@ namespace JME {
             static const JetResolutionScaleFactor get(const edm::EventSetup&, const std::string&);
 #endif
 
-            float getScaleFactor(const JetParameters& parameters, Variation variation = Variation::NOMINAL) const;
+    float getScaleFactor(const JetParameters& parameters,
```
Comment (slava77) [inline JetMETCorrections/Modules/interface/JetResolution.h]: indentation

File: JetMETCorrections/Modules/src/JetResolution.cc [hunk 4]
```diff
@@ -53,13 +53,32 @@ namespace JME {
...
+                                                 Variation variation /* = Variation::NOMINAL*/,
+                                                 std::string uncertaintySource /* = ""*/) const {
         const JetResolutionObject::Record* record = m_object->getRecord(parameters);
         if (! record)
             return 1;
 
         const std::vector<float>& parameters_values = record->getParametersValues();
-        return parameters_values[static_cast<size_t>(variation)];
+    const std::vector<std::string>& parameter_names = m_object->getDefinition().getParametersName();
+    size_t parameter = static_cast<size_t>(variation);
```
Comment (slava77) [inline JetMETCorrections/Modules/src/JetResolution.cc]: indentation

**Review by slava77 — COMMENTED** [review]
there is the same issue as in the 106X backport related to the code indentation in most of the files

File: JetMETCorrections/Modules/src/JetResolution.cc (hunk 4 above)
Comment (ahinzmann) [reply JetMETCorrections/Modules/src/JetResolution.cc]: done

File: JetMETCorrections/Modules/interface/JetResolution.h (hunk 3 above)
Comment (ahinzmann) [reply JetMETCorrections/Modules/interface/JetResolution.h]: done

File: CondFormats/JetMETObjects/src/JetResolutionObject.cc (hunk 2 above)
Comment (ahinzmann) [reply CondFormats/JetMETObjects/src/JetResolutionObject.cc]: done

File: CondFormats/JetMETObjects/interface/JetResolutionObject.h (hunk 1 above)
Comment (ahinzmann) [reply CondFormats/JetMETObjects/interface/JetResolutionObject.h]: done

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/28518	860f3bc
- code changes are in line with the PR description and the follow up review
    - from #28480 the new code is not enabled in available standard configuration(s), to be used first at the analysis or special nanoAOD production level
    - the backport looks correct
- jenkins tests pass and comparisons with the baseline show no differences

**Comment by ggovi** [comment]
+1

**Comment by santocch** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**